src/consensus/produce_block.rs

```diff
@@ -116,28 +117,58 @@ where
     pub(crate) async fn produce_block(
         &self,
         slot: Slot,
-        parent: (Slot, Hash),
-        parent_ready: bool,
+        parent: BlockId,
+        parent_ready: &mut Option<oneshot::Receiver<BlockId>>,
     ) -> Result<()> {
         let (parent_slot, parent_hash) = parent;
+        if parent_ready.is_some() {
+            assert!(slot.is_start_of_window());
+        }
         let _slot_span = Span::enter_with_local_parent(format!("slot {slot}"));
         info!(
             "producing block in slot {} with parent {} in slot {} (parent ready: {})",
             slot,
             &hex::encode(parent_hash)[..8],
             parent_slot,
-            parent_ready,
+            parent_ready.is_none(),
         );
 
         let mut sleep_duration = DELTA_BLOCK;
+        let mut parent_changed = None;
```

Comment:
bug: not handling scenario where we see a parent ready later but it is for the same parent that we have building on top of so far.  We should not produce a slice with a new parent set then.